Fix deprecation messages for `bundle install` flags, the config should be --local as before

eregon opened the pull request:
* Closes https://github.com/rubygems/rubygems/issues/3916.
* Always show --local or --global for `bundle config set` commands as
  the default is global which can be not intuitive (#3916).

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

deivid-rodriguez commented on bundler/spec/commands/outdated_spec.rb:
```diff
@@ -555,15 +555,15 @@ def test_group_option(group)
         gem "foo"
       G
-      bundle "config set deployment true"
+      bundle "config --local deployment true"
```
Could you keep `set`?

eregon replied:
Done

deivid-rodriguez commented on bundler/lib/bundler/fetcher.rb:
```diff
@@ -47,7 +47,7 @@ def initialize(remote_uri)
         super "Authentication is required for #{remote_uri}.\n" \
           "Please supply credentials for this source. You can do this by running:\n" \
-          " bundle config set #{remote_uri} username:password"
+          " bundle config set --global #{remote_uri} username:password"
```
In the other places, you chose `--local` for configuring credentials for a source. Should we use that here too?

eregon replied:
Done, but might change again for https://github.com/rubygems/rubygems/pull/3917#discussion_r479780099

deivid-rodriguez replied:
Yeah, I saw your other comment after I posted this.

eregon commented on bundler/man/bundle-config.ronn:
```diff
@@ -333,21 +333,21 @@ Bundler supports overriding gem sources with mirrors. This allows you to
 mirror to fetch gems.
 
-    bundle config set mirror.SOURCE_URL MIRROR_URL
+    bundle config set --local mirror.SOURCE_URL MIRROR_URL
```
Not sure if better to show `--local` or `--global` for mirror and authentication (just below).
But at least it seems better to show one of them, so users can consider whether they want it local or global and not rely on the default which might not do the right thing.

deivid-rodriguez replied:
Agreed. Since we are unsure, let's go with `--global` I guess to match current "flagless" behaviour?

eregon replied:
OK, switched to `--global` for mirror & auth config commands.

deivid-rodriguez reviewed (commented):
Looks good! Being explicit is a good step forward for now, and makes it easier to change the default in the future.

eregon commented on bundler/man/bundle-config.1.txt:
```diff
@@ -70,43 +70,44 @@ REMEMBERING OPTIONS
+              environment's  PATH  variable.  For  instance,  if the rails gem
               comes with a rails executable, this flag will create a bin/rails
-              executable that ensures that all referred dependencies  will  be
+              executable  that  ensures that all referred dependencies will be
```
My `groff` seems to generate unnecessary changes like here, could someone run `rake man:build` with a version that doesn't cause so many changes? Then I'll pick it up in this PR.

deivid-rodriguez replied:
Yes, I hate it :angry:. My `groff` version does that too. I think we might want to pass a ruby script over the result of `groff` (instead of just `col -xb`) to eliminate these whitespace differences.

eregon replied:
I made https://github.com/eregon/rubygems/commit/ad68c24e09e812b906dbe4a1542ac33f6d7482e8 that fixes the extra `1m` and `22m` in the diff:
```diff
-NAME
-       bundle-add - Add gem to the Gemfile and run bundle install
+1mNAME0m
+       1mbundle-add 22m- Add gem to the Gemfile and run bundle install
-SYNOPSIS
-       bundle     add     GEM_NAME     [--group=GROUP]     [--version=VERSION]
+1mSYNOPSIS0m
+       1mbundle     add     4m22mGEM_NAME24m     [--group=GROUP]     [--version=VERSION]
        [--source=SOURCE]   [--git=GIT]   [--branch=BRANCH]    [--skip-install]
```
but it does not fix those whitespace changes. So I wonder if `groff` somehow forgot to strip those escape sequences and that causes the extra spacing changes.

eregon replied:
BTW, what are the `.txt` for? Isn't the `.ronn` enough for "readable without `man`"?

deivid-rodriguez replied:
> but it does not fix those whitespace changes. So I wonder if groff somehow forgot to strip those escape sequences and that causes the extra spacing changes.

I think the extra spacing is generated due to `ronn` generating TABs at random places when you make changes to a ronn source. But not really sure.

> BTW, what are the .txt for? Isn't the .ronn enough for "readable without man"?

That's a great point. We can probably print the `ronn` sources when `man` is not available and stop generating these txt files.

eregon replied:
I ended up using `git add -p`, and looking at which lines were relevant with `git diff -w`. Quite bothersome but it worked.

eregon commented:
@deivid-rodriguez I think I addressed everything now, and I added a CHANGELOG entry, should be good to go, except that regeneration of `bundle-config.1.txt` (unless we're fine with it?).

eregon commented:
FWIW I like `bundle config --local path vendor/bundle` better than `bundle config set --local path vendor/bundle` but given that seems deprecated for Bundler 3 I went with the Bundler 3 syntax and explicit `set` (which is another difference to `git config`).
(Specifically if we change the default to `--local`, `bundle config path vendor/bundle` sounds nicer than `bundle config set path vendor/bundle` to me)

deivid-rodriguez commented:
> @deivid-rodriguez I think I addressed everything now, and I added a CHANGELOG entry, should be good to go, except that regeneration of bundle-config.1.txt (unless we're fine with it?).

You can skip the changelog message since I plan to start generating the changelog automatically from merge PR titles and labels in the next version.

> FWIW I like `bundle config --local path vendor/bundle` better than `bundle config set --local path vendor/bundle` but given that seems deprecated for Bundler 3 I went with the Bundler 3 syntax and explicit set (which is another difference to `git config`).
(Specifically if we change the default to `--local`, `bundle config path vendor/bundle` sounds nicer than `bundle config set path vendor/bundle` to me)

I'm also unsure about deprecating the old `bundle config` interface, that's why I delayed it until it can be better discussed, and the original reasons properly reviewed: https://github.com/rubygems/bundler/pull/7475. But we're using the new interface internally in a consistent way at the moment, so let's keep doing it for the moment :+1:.

deivid-rodriguez commented:
Also, I think we can ignore the annoying txt diff for this PR, and try to get rid of txt files at all as you suggested.

eregon commented:
> You can skip the changelog message since I plan to start generating the changelog automatically from merge PR titles and labels in the next version.

Should I remove that change then?

> I'm also unsure about deprecating the old `bundle config` interface, that's why I delayed it until it can be better discussed, and the original reasons properly reviewed: [rubygems/bundler#7475](https://github.com/rubygems/bundler/pull/7475). But we're using the new interface internally in a consistent way at the moment, so let's keep doing it for the moment .

Is there some good place to discuss that? I feel this deprecation is going to be rather annoying (and require verbose commands) for very little gains. The only conflict would be if one needs to set a key named `list/get/set/unset` but that seems extremely unlikely.

> Also, I think we can ignore the annoying txt diff for this PR, and try to get rid of txt files at all as you suggested.

Will you make a PR or is there a plan for that then? :smiley: 
I guess we should wait to merge this PR after that?

I think this PR should be merged for the next Bundler 2 release, it's quite bad if people configure globally when they do not intend to (can cause confusion, or an impression that the deprecation messages are just broken and due to that might be ignored more).

deivid-rodriguez commented:
> Should I remove that change then?

It will be overwritten unless the PR title matches the current wording exactly, so yeah I'd rather remove it!

> Is there some good place to discuss that? I feel this deprecation is going to be rather annoying (and require verbose commands) for very little gains. The only conflict would be if one needs to set a key named list/get/set/unset but that seems extremely unlikely.

Feel free to open a new ticket, investigate why the new interface was added, and explain why you think the reasons for making it compulsory are wrong.

> Will you make a PR or is there a plan for that then? :smiley:
> I guess we should wait to merge this PR after that?

I like your idea, but it came up literally 1 hour ago, so I'm not sure when I (or anybody) will get to it. Better not to block this PR on that.

> I think this PR should be merged for the next Bundler 2 release, it's quite bad if people configure globally when they do not intend to (can cause confusion, or an impression that the deprecation messages are just broken and due to that might be ignored more).

Sure, just be patient! :)
Sa little
This is straightforward and will make it to the next release, just nbe

eregon commented:
OK, CHANGELOG entry removed. I worked around and now the diff for `bundle-config.1.txt` should be minimal.
Ready to be merged :)

deivid-rodriguez commented:
Thanks!
